cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -0,0 +1,123 @@
+
+
+def _get_dd_sequence_from_schema_name(schema: str) -> Sequence['cirq.Gate']:
+    """Gets dynamical decoupling sequence from a schema name."""
+    dd_sequence: Sequence['cirq.Gate']
+    match schema:
+        case 'XX_PAIR':
+            dd_sequence = (cirq.X, cirq.X)
+        case 'X_XINV':
+            dd_sequence = (cirq.X, cirq.X**-1)
+        case 'YY_PAIR':
+            dd_sequence = (cirq.Y, cirq.Y)
+        case 'Y_YINV':
+            dd_sequence = (cirq.Y, cirq.Y**-1)
+        case _:
+            raise ValueError('Invalid schema name.')
+    return dd_sequence
+
+
+def _validate_dd_sequence(dd_sequence: Sequence['cirq.Gate']) -> Tuple[bool, Optional[str]]:
+    """Validates a given dynamical decoupling sequence.
+
+    Args:
+        dd_sequence: Input dynamical sequence to be validated.
+
+    Returns:
+        A tuple containing:
+            - is_valid (bool): True if the dd sequence is valid, False otherwise.
+            - error_message (str): An error message if the dd sequence is invalid, else None.
+    """
+    if len(dd_sequence) < 2:
+        return False, 'Invalid dynamical decoupling sequence. Expect more than one gates.'
+    matrices = [cirq.unitary(gate) for gate in dd_sequence]
+    product = reduce(np.matmul, matrices)
+
+    if not cirq.equal_up_to_global_phase(product, np.eye(2)):
+        return False, (
+            'Invalid dynamical decoupling sequence. Expect sequence production equals'
+            f' identity up to a global phase, got {product}.'.replace('\n', ' ')
+        )
+    return True, None
+
+
+@transformer_api.transformer
+def add_dynamical_decoupling(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    schema: Union[str, Sequence['cirq.Gate']] = 'X_XINV',
+) -> 'cirq.Circuit':
+    """Adds dynamical decoupling gate operations to idle moments of a given circuit.
+    This transformer preserves the moment structure of the circuit.
+
+    Args:
+          circuit: Input circuit to transform.
+          context: `cirq.TransformerContext` storing common configurable options for transformers.
+          schema: Dynamical decoupling schema name or a dynamical decoupling sequence.
+            If a schema is specified, provided dynamical decouping sequence will be used.
+            Otherwise, customized dynamical decoupling sequence will be applied.
+
+    Returns:
+          A copy of the input circuit with dynamical decoupling operations.
+
+    Raises:
+          ValueError: If schema is not valid.
+    """
+    last_busy_moment_by_qubits: Dict['cirq.Qid', int] = {q: 0 for q in circuit.all_qubits()}
+    insert_into: list[Tuple[int, 'cirq.OP_TREE']] = []
+
+    if isinstance(schema, str):
+        try:
+            base_dd_sequence = _get_dd_sequence_from_schema_name(schema)
+        except ValueError:
+            raise
+    else:
+        is_valid, error_message = _validate_dd_sequence(schema)
```

Comment:
this is a golang style, prefer to just raise the error inside `_validate_dd_sequence`

Reply:
Done.